### PR TITLE
improvements to metadata sync and data sync tasks

### PR DESCRIPTION
**Pull request by sultanahamer**
Data sync task
 - Events sync summary is now shown when events post api call fails with 409
 - handles  HTTPClient/Server Exception and resourceAccessExceptions
Metadata sync task
 - Importer fails both runtime and validation are handled
 - retries when download of version's data(/api/metadata/version/Version_1/data.gz) fails.
Metadata sync apis
 - create api now returns valid response upon issues.
Misc
 - Implemented a generic WebMessageParser

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/sync/exception/MetadataSyncImportException.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package org.hisp.dhis.dxf2.metadata.sync.exception;
+
+import net.sf.jasperreports.engine.util.JRStyledText;
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/sync/exception/MetadataSyncImportException.java]: Unnecessary import.

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/tasks/MetadataRetryContext.java [hunk 2]
```diff
@@ -50,6 +51,7 @@
     private static final Log log = LogFactory.getLog( MetadataRetryContext.class );
 
     private RetryContext retryContext;
+    private String[] AttributeAccessor;
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/tasks/MetadataRetryContext.java]: Use lowercase start of variable names.

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/synch/DefaultSynchronizationManager.java [hunk 3]
```diff
@@ -243,23 +245,38 @@ private ImportSummary executeDataPush( SystemInstance instance )
...
+            String responseBody = ex.getResponseBodyAsString();
+            summary = DefaultWebMessageJacksonService.fromWebMessageResponse( responseBody, ImportSummary.class );
+        }
+        catch (HttpServerErrorException ex)
+        {
+            String responseBody = ex.getResponseBodyAsString();
+            log.error( "Internal error happened during event data push: " + responseBody , ex);
+            throw ex;
+        }
+        catch (ResourceAccessException ex)
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/synch/DefaultSynchronizationManager.java]: Codestyle, space after/before parenthesis.

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/webmessage/DefaultWebMessageJacksonService.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.StringWriter;
+import java.util.Date;
+
+/**
+ * Created by vanyas on 5/4/17.
+ */
+public class DefaultWebMessageJacksonService
+{
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/webmessage/DefaultWebMessageJacksonService.java]: This seems more of a utility class than a service. Can we move these methods to the existing JacksonUtils class in dhis-support-system project?

Also the XML mapper does not seem to be in use, could be removed?

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/CrudControllerAdvice.java [hunk 5]
```diff
@@ -227,6 +230,27 @@ public void handleDhisVersionMismatchException( DhisVersionMismatchException ver
         webMessageService.send( WebMessageUtils.forbidden( versionMismatchException.getMessage() ), response, request );
     }
 
+    @ExceptionHandler( MetadataImportConflictException.class )
+    public void handleMetadataImportConflictException( MetadataImportConflictException conflictException, HttpServletResponse response, HttpServletRequest request )
+    {
+
+        if(conflictException.getMetadataSyncSummary() == null)
+            webMessageService.send( WebMessageUtils.conflict( conflictException.getMessage() ), response, request );
```
Comment (larshelge) [inline dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/CrudControllerAdvice.java]: Codesyle, use brackets and spaces

**Comment by ceciliapersson** [comment]
FYI, I've recently reviewed the whole section "Configure metadata synchronization" in the DHIS2 user guide (https://docs.dhis2.org/master/en/user/html/metadata_sync.html).

Please let me know if any additional updates are required due to this pull request.

**Comment by sultanahamer** [comment]
@ceciliapersson will we go through the documentation and let you know any changes if required by tomorrow.
@larshelge will make changes according to comments by tomorrow.

**Comment by vanyaseth** [comment]
@ceciliapersson 
The user manual documentation looks fine. There is no change there. 
This PR will result in a change in the developer manual. We will update that by next week.

**Comment by ceciliapersson** [comment]
@vanyaseth Thanks for the information!

**Comment by sultanahamer** [comment]
@larshelge Made changes according to comments, except for moving DefaultWebMessageJacksonService.java  to utils in support system. Moving that requires support system depend on core. Hence we have implemented a new utils class in core.

**Comment by larshelge** [comment]
Thanks for the fixes. Ready to merge.

**Comment by vanyaseth** [comment]
hi Lars,

These changes need to be in 2.26 as well. Do you want us to send a PR for that? 
Or can you take care of the merge?

Regards
Vanya

**Pull request merged**